sa/model_test.go

```diff
@@ -224,3 +239,144 @@ func TestPopulateAttemptedFieldsBadJSON(t *testing.T) {
 		})
 	}
 }
+
+func TestCerficatesTableContainsDuplicateSerials(t *testing.T) {
+	testCtx := setup(t)
+	defer testCtx.cleanUp()
+
+	serialString := core.SerialToString(big.NewInt(1337))
+
+	// Register and insert a certificate with a serial of `1337`.
+	testCtx.setupCertificate1337A(t)
+
+	// This should return the certificate that we just inserted.
+	_, err := SelectCertificate(testCtx.db, serialString)
+	test.AssertNotError(t, err, "received an error for a valid query")
+
+	// Register and insert a certificate with a serial of `1337` but for a
+	// different hostname.
+	testCtx.setupCertificate1337B(t)
+
+	// With a duplicate being present, this should error.
+	_, err = SelectCertificate(testCtx.db, serialString)
+	test.AssertError(t, err, "should've received an error for multiple rows")
+}
+
+type testCtx struct {
+	db      *db.WrappedMap
+	ssa     core.StorageAdder
```

Comment:
Let's use the concrete type `*SQLStorageAuthority` here. I don't see any need to abstract via the interface.